### ISSUE TITLE
Add support for OSC 22

### ISSUE THREAD

**Issue by kchibisov** [alacritty/alacritty#7119]
This allows setting a cursor shape, the names are X11 names which should be just
passed as is to winit. We may want to have an option to disable it.

The use case is to have a better mouse mode support in some applications.

The escape is supported by: foot, xterm. The docs are on https://invisible-island.net/xterm/ctlseqs/ctlseqs.html . If the shape doesn't exist we can ignore the event.

**Comment by chrisduerr** [comment]
> The use case is to have a better mouse mode support in some applications.

How does this improve mouse mode support? I'd argue that this makes mouse mode support much worse, because now you can't be sure which mode you're in anymore.

I'm generally not a huge fan of window ops and this one seems to have some obvious drawbacks?

**Comment by kchibisov** [comment]
@chrisduerr in general it's nice to have a resize cursor for tmux panes, loading cursor when you have some action running to provide an aditional indication. it's true that you can't judge by default/text cursor anymore, but only if the application sets the text cursor itself, though most can select.

Like it basically provides a visual indication that you can do something which you won't show otherwise with any reasonable mechanism. This option is opt-in in terminal applications in general, so it's not like it'll heart, but I'll use it myself.

**Comment by rockorager** [comment]
As an application developer, there are few downsides I see to this one. Terminals either ignore it, or respect it. Users of those respecting OSC22 get better hinting at usability. The few shapes I really care about are resize shapes, cell shapes, clickable, and busy/backround busy shapes. All of these provide huge UI improvements when using a mouse in a TUI.

**Comment by chrisduerr** [comment]
I'm definitely not sold at all, so at the very least it would require an option to disable this escape.

I'd assume there's no capability for this?

**Comment by rockorager** [comment]
@chrisduerr There is no query for this one, nor any terminfo cap that I'm aware of. Here's a quick [demo](https://cdn.fosstodon.org/media_attachments/files/110/793/406/955/803/781/original/9ddf11a8243c70d3.mp4).

**Comment by j4james** [comment]
Just FYI, if you care about cross-platform support, this sequence isn't great. On XTerm the cursor styles come directly from [Xlib](https://x.org/releases/current/doc/libX11/libX11/libX11.html#x_font_cursors), but the only other terminal I know of that supports `OSC 22` is Mintty, whose implementation is based on the [Windows cursor names](https://learn.microsoft.com/en-us/windows/win32/menurc/about-cursors).

And some of the X styles are just bizarre: things like "bogosity", "gumby", and "gobbler". It's like a 90's Geocities web page. The Windows styles are more reasonably thought out, but there's very little overlap between the two. The only names in common are "arrow" and "cross", although technically "ibeam" also works, because that's the fallback in both terminals.

This is something that could definitely do with standardization, even if it's just an agreed upon subset of common names.

**Comment by kchibisov** [comment]
w3c css cursors names must be used, everything else in such case is nonsense.

**Comment by j4james** [comment]
> w3c css cursors names must be used, everything else in such case is nonsense.

Yeah. That make perfect sense to me. I've just been looking at the [foot implementation](https://codeberg.org/dnkl/foot/src/branch/master/cursor-shape.c#L62), and it appears that's actually what they're doing. Although it seems like they also include some aliases for compatibility with the X names.

**Comment by kchibisov** [comment]
That's because they are Wayland terminal and cursor-shape protocol uses w3c cursors as well.
